File: iOS/SayHi/SayHi/Enum.swift
//
//  Enum.swift
//  SayHi
//
//  Created by Klemenz, Oliver on 07.04.17.
//  Copyright © 2017 Klemenz, Oliver. All rights reserved.
//

import Foundation

enum Gender : String {
    case none   = ""
    case male   = "m"
    case female = "f"
    
    static let list : [Gender] = [.none, .male, .female]
}

enum MatchMode : Int {
    case basic = 1
    case exact = 2
    case adapt = 3
    case tries = 4
    case open  = 5
    
    var description: String {
        switch self {
        case .basic:
            return "basic"
        case .exact:
            return "exact"
        case .adapt:
            return "adapt"
        case .tries:
            return "tries"
        case .open:
            return "open"
        }
    }
    
    var externalDescription: String {
        switch self {
        case .basic:
            return "basic"
        case .exact:
            return "exact"
        case .adapt:
            return "adapt"
        case .tries:
            return "try"
        case .open:
            return "open"
        }
    }
    
    var numberedDescription: String {
        return "\(description)_\(rawValue)"
    }
    
    static func fromDescription(_ description: String) -> MatchMode? {
        if description == "basic" {
            return .basic
        } else if description == "exact" {
            return .exact
        } else if description == "adapt" {
            return .adapt
        } else if description == "tries" {
            return .tries
        } else if description == "open" {
            return .open
        }
        return nil
    }
    
    static let list : [MatchMode] = [.basic, .exact, .adapt, .tries, .open]
}

enum PasscodeTimeout : Int {
    case min0  = 0
    case min1  = 1
    case min5  = 5
    case min10 = 10
    
    static let list : [PasscodeTimeout] = [.min0, .min1, .min5, .min10]
}

enum RelationType : String {
    
    case none = "None"
    case date = "Date"
    case single = "Single"
    case family = "Family"
    case partner = "Partner"
    case wife = "Wife"
    case husband = "Husband"
    case parent = "Parent"
    case father = "Father"
    case mother = "Mother"
    case child = "Child"
    case son = "Son"
    case daugther = "Daugther"
    case sibling = "Sibling"
    case brother = "Brother"
    case sister = "Sister"
    case grandparent = "Grandparent"
    case grandfather = "Grandfather"
    case grandmother = "Grandmother"
    case grandchild = "Grandchild"
    case grandson = "Grandson"
    case granddaughter = "Granddaughter"
    case relative = "Relative"
    case uncle = "Uncle"
    case aunt = "Aunt"
    case cousins = "Cousins"
    case cousin = "Cousin (m)"
    case cousine = "Cousin (f)"
    case nephew = "Nephew"
    case niece = "Niece"
    case friend = "Friend"
    case boyfriend = "Boyfriend"
    case girlfriend = "Girlfriend"
    case colleague = "Colleague"
    case boss = "Boss"
    case manager = "Manager"
    case employee = "Employee"
    case peer = "Peer"
    case other = "Other"
    
    static let list : [RelationType] = [.none, .date, .single,
                                        .boyfriend, .girlfriend, .wife, .husband,
                                        .partner, .family, .friend,
                                        .child, .son, .daugther,
                                        .parent, .father, .mother,
                                        .sibling, .brother, .sister,
                                        .relative, .uncle, .aunt,
                                        .grandchild, .grandson, .granddaughter,
                                        .grandparent, .grandfather, .grandmother,
                                        .cousins, .cousin, .cousine,
                                        .nephew, .niece,
                                        .colleague, .boss, .manager, .employee, .peer,
                                        .other]
}

File: Android/SayHi/app/src/main/java/de/oklemenz/sayhi/model/Enum.java
package de.oklemenz.sayhi.model;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;

import de.oklemenz.sayhi.service.Utilities;

/**
 * Created by Oliver Klemenz on 31.10.16.
 */

public class Enum implements Serializable {

    public enum Gender {
        None(""),
        Male("m"),
        Female("f");

        public String code;

        public static Gender fromCode(String code) {
            switch (code) {
                case "":
                    return None;
                case "m":
                    return Male;
                case "f":
                    return Female;
            }
            return None;
        }

        Gender(String code) {
            this.code = code;
        }

        public static List<Gender> list = Arrays.asList(None, Male, Female);
    }

    public enum MatchMode {
        Basic(1),
        Exact(2),
        Adapt(3),
        Tries(4),
        Open(5);

        public int code;

        public String toString() {
            switch (code) {
                case 1:
                    return "basic";
                case 2:
                    return "exact";
                case 3:
                    return "adapt";
                case 4:
                    return "tries";
                case 5:
                    return "open";
            }
            return "exact";
        }

        public String toExternalString() {
            switch (code) {
                case 1:
                    return "basic";
                case 2:
                    return "exact";
                case 3:
                    return "adapt";
                case 4:
                    return "try";
                case 5:
                    return "open";
            }
            return "exact";
        }

        public String toNumberedString() {
            return toString() + "_" + code;
        }

        public static MatchMode fromCode(int code) {
            switch (code) {
                case 1:
                    return Basic;
                case 2:
                    return Exact;
                case 3:
                    return Adapt;
                case 4:
                    return Tries;
                case 5:
                    return Open;
            }
            return Exact;
        }

        public static MatchMode fromExternalCode(String code) {
            switch (code) {
                case "basic":
                    return Basic;
                case "exact":
                    return Exact;
                case "adapt":
                    return Adapt;
                case "try":
                    return Tries;
                case "open":
                    return Open;
            }
            return Exact;
        }

        MatchMode(int code) {
            this.code = code;
        }

        public static List<MatchMode> list = Arrays.asList(Basic, Exact, Adapt, Tries, Open);
    }

    public enum PasscodeTimeout {
        Min0(0),
        Min1(1),
        Min5(5),
        Min10(10);

        public int code;

        public String toString() {
            switch (code) {
                case 0:
                    return "min0";
                case 1:
                    return "min1";
                case 5:
                    return "min5";
                case 10:
                    return "min10";
            }
            return "min5";
        }

        public static PasscodeTimeout fromCode(int code) {
            switch (code) {
                case 0:
                    return Min0;
                case 1:
                    return Min1;
                case 5:
                    return Min5;
                case 10:
                    return Min10;
            }
            return Min0;
        }

        PasscodeTimeout(int code) {
            this.code = code;
        }

        public static List<PasscodeTimeout> list = Arrays.asList(Min0, Min1, Min5, Min10);
    }

    public enum CorrectionLevel {
        Low("L"),
        Medium("M"),
        Quartile("Q"),
        High("H");

        public String code;

        public static CorrectionLevel fromCode(String code) {
            switch (code) {
                case "L":
                    return Low;
                case "M":
                    return Medium;
                case "Q":
                    return Quartile;
                case "H":
                    return High;
            }
            return Low;
        }

        CorrectionLevel(String code) {
            this.code = code;
        }

        public static List<CorrectionLevel> list = Arrays.asList(Low, Medium, Quartile, High);
    }

    public enum RelationType {
        None("none"),
        Date("date"),
        Single("single"),
        Family("family"),
        Partner("partner"),
        Wife("wife"),
        Husband("husband"),
        Parent("parent"),
        Father("father"),
        Mother("mother"),
        Child("child"),
        Son("son"),
        Daugther("daugther"),
        Sibling("sibling"),
        Brother("brother"),
        Sister("sister"),
        Grandparent("grandparent"),
        Grandfather("grandfather"),
        Grandmother("grandmother"),
        Grandchild("grandchild"),
        Grandson("grandson"),
        Granddaughter("granddaughter"),
        Relative("relative"),
        Uncle("uncle"),
        Aunt("aunt"),
        Cousins("cousins"),
        Cousin("cousin"),
        Cousine("cousine"),
        Nephew("nephew"),
        Niece("niece"),
        Friend("friend"),
        Boyfriend("boyfriend"),
        Girlfriend("girlfriend"),
        Colleague("colleague"),
        Boss("boss"),
        Manager("manager"),
        Employee("employee"),
        Peer("peer"),
        Other("other");

        public String code;

        public String toString() {
            return Utilities.uppercaseFirst(code);
        }

        public static RelationType fromCode(String code) {
            switch (code) {
                case "":
                    return None;
                case "date":
                    return Date;
                case "single":
                    return Single;
                case "family":
                    return Family;
                case "partner":
                    return Partner;
                case "wife":
                    return Wife;
                case "husband":
                    return Husband;
                case "parent":
                    return Parent;
                case "father":
                    return Father;
                case "mother":
                    return Mother;
                case "child":
                    return Child;
                case "son":
                    return Son;
                case "daugther":
                    return Daugther;
                case "sibling":
                    return Sibling;
                case "brother":
                    return Brother;
                case "sister":
                    return Sister;
                case "grandparent":
                    return Grandparent;
                case "grandfather":
                    return Grandfather;
                case "grandmother":
                    return Grandmother;
                case "grandchild":
                    return Grandchild;
                case "grandson":
                    return Grandson;
                case "granddaughter":
                    return Granddaughter;
                case "relative":
                    return Relative;
                case "uncle":
                    return Uncle;
                case "aunt":
                    return Aunt;
                case "cousins":
                    return Cousins;
                case "cousin":
                    return Cousin;
                case "cousine":
                    return Cousine;
                case "nephew":
                    return Nephew;
                case "niece":
                    return Niece;
                case "friend":
                    return Friend;
                case "boyfriend":
                    return Boyfriend;
                case "girlfriend":
                    return Girlfriend;
                case "colleague":
                    return Colleague;
                case "boss":
                    return Boss;
                case "manager":
                    return Manager;
                case "employee":
                    return Employee;
                case "peer":
                    return Peer;
                case "other":
                    return Other;
            }
            return None;
        }

        RelationType(String code) {
            this.code = code;
        }

        public static RelationType fromDescription(String description) {
            return Enum.RelationType.fromCode(Utilities.lowercaseFirst(description));
        }

        public static List<RelationType> list = Arrays.asList(
                None, Date, Single,
                Boyfriend, Girlfriend, Wife, Husband,
                Partner, Family, Friend,
                Child, Son, Daugther,
                Parent, Father, Mother,
                Sibling, Brother, Sister,
                Relative, Uncle, Aunt,
                Grandchild, Grandson, Granddaughter,
                Grandparent, Grandfather, Grandmother,
                Cousins, Cousin, Cousine,
                Nephew, Niece,
                Colleague, Boss, Manager, Employee, Peer,
                Other);
    }
}
